src/Storages/ObjectStorage/StorageObjectStorageCluster.cpp

```diff
@@ -0,0 +1,100 @@
+
+#include <Common/Exception.h>
+#include <Parsers/queryToString.h>
+#include <Processors/Sources/RemoteSource.h>
+#include <QueryPipeline/RemoteQueryExecutor.h>
+
+#include <Storages/VirtualColumnUtils.h>
+#include <Storages/ObjectStorage/Utils.h>
+#include <Storages/ObjectStorage/StorageObjectStorageSource.h>
+#include <Storages/extractTableFunctionArgumentsFromSelectQuery.h>
+
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int LOGICAL_ERROR;
+}
+
+StorageObjectStorageCluster::StorageObjectStorageCluster(
+    const String & cluster_name_,
+    ConfigurationPtr configuration_,
+    ObjectStoragePtr object_storage_,
+    const StorageID & table_id_,
+    const ColumnsDescription & columns_,
+    const ConstraintsDescription & constraints_,
+    ContextPtr context_)
+    : IStorageCluster(
+        cluster_name_, table_id_, getLogger(fmt::format("{}({})", configuration_->getEngineName(), table_id_.table_name)))
+    , configuration{configuration_}
+    , object_storage(object_storage_)
+{
+    ColumnsDescription columns{columns_};
+    resolveSchemaAndFormat(columns, configuration->format, object_storage, configuration, {}, context_);
+    configuration->check(context_);
+
+    StorageInMemoryMetadata metadata;
+    metadata.setColumns(columns);
+    metadata.setConstraints(constraints_);
+
+    setVirtuals(VirtualColumnUtils::getVirtualsForFileLikeStorage(metadata.getColumns()));
+    setInMemoryMetadata(metadata);
+}
+
+std::string StorageObjectStorageCluster::getName() const
+{
+    return configuration->getEngineName();
+}
+
+void StorageObjectStorageCluster::updateQueryToSendIfNeeded(
+    ASTPtr & query,
+    const DB::StorageSnapshotPtr & storage_snapshot,
+    const ContextPtr & context)
+{
+    ASTExpressionList * expression_list = extractTableFunctionArgumentsFromSelectQuery(query);
+    if (!expression_list)
+    {
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "Expected SELECT query from table function {}, got '{}'",
+            configuration->getEngineName(), queryToString(query));
+    }
+
+    ASTs & args = expression_list->children;
+    const auto & structure = storage_snapshot->metadata->getColumns().getAll().toNamesAndTypesDescription();
+    if (args.empty())
+    {
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR,
+            "Unexpected empty list of arguments for {}Cluster table function",
+            configuration->getEngineName());
+    }
+
+    ASTPtr cluster_name_arg = args.front();
+    args.erase(args.begin());
+    configuration->addStructureAndFormatToArgs(args, structure, configuration->format, context);
+    args.insert(args.begin(), cluster_name_arg);
+}
+
+RemoteQueryExecutor::Extension StorageObjectStorageCluster::getTaskIteratorExtension(
+    const ActionsDAG::Node * predicate, const ContextPtr & local_context) const
+{
+    auto iterator = StorageObjectStorageSource::createFileIterator(
```

Comment:
Interesting how `s3Cluster` works with archives. I guess it doesn't work correctly

Reply:
> Interesting how s3Cluster works with archives. Maybe need a test for it

I was planning to ask Daniil who made this feature to make a test for this case because when applying his changes to this PR - I also had a thought that it does not work with s3Cluster, but I wanted to ask him to test it after I merge this PR so that he does not add me more conflicts :)